src/chapters/control-flow-and-collections/hashmap.rst

```diff
@@ -0,0 +1,228 @@
+.. index:: ! HashMap, ! map
+
+HashMap
+=======
+
+Java also provides us a structure to store
+data as key/value pairs. Java calls these objects **hashmaps** 
+(or **maps**, more generally), and they are provided by the ``HashMap`` class.
+
+Considering the gradebook example, we can improve our program using a
+map. We'll store the students’ grades along with their names in the same
+data structure. The names will be the keys, and the grades will be the
+values.
+       
+As with the other collection structures, in Java we must specify the types of 
+the objects we’ll be storing when we declare a variable or parameter to be a map.
+This means specifying both key and value data types, which are allowed
+to be different types for a given map.
+
+.. sourcecode:: java
+   :linenos:
+
+   package org.launchcode.java.demos.collections;
+
+   import java.util.HashMap;
+   import java.util.Map;
+   import java.util.Scanner;
+   
+   public class HashMapGradebook {
+
+      public static void main(String[] args) {
+
+         HashMap<String, Double> students = new HashMap<>();
+         Scanner in = new Scanner(System.in);
+         String newStudent;
+
+         System.out.println("Enter your students (or ENTER to finish):");
+
+         // Get student names and grades
+         do {
+
+            System.out.print("Student: ");
+            newStudent = in.nextLine();
+
+            if (!newStudent.equals("")) {
+               System.out.print("Grade: ");
+               Double newGrade = in.nextDouble();
+               students.put(newStudent, newGrade);
+
+               // Read in the newline before looping back
+               in.nextLine();
+            }
+
+         } while(!newStudent.equals(""));
+
+         // Print class roster
+         System.out.println("\nClass roster:");
+         double sum = 0.0;
+
+         for (Map.Entry<String, Double> student : students.entrySet()) {
+            System.out.println(student.getKey() + " (" + student.getValue() + ")");
+            sum += student.getValue();
+         }
+
+         double avg = sum / students.size();
+         System.out.println("Average grade: " + avg);
+      }    
+   }
+
+We can add a new item with a ``.put()`` method, specifying both key and
```

Comment:
How about a short explanation here for line 11 - which sets the types for <key,value> - and line 26 (``put``).